File: shelf/entity.go
package shelf

import (
	"time"
)

type Entity interface {
	ID() string
	Date() time.Time
	MimeType() string
	Description() string
	Path() string
}

type entity struct {
	ID_          string    `yaml:"-"`
	Path_        string    `yaml:"-"`
	Date_        time.Time `yaml:"date"`
	MimeType_    string    `yaml:"mime-type"`
	Description_ string    `yaml:"description"`
}

func (e *entity) ID() string {
	return e.ID_
}
func (e *entity) Date() time.Time {
	return e.Date_
}
func (e *entity) MimeType() string {
	return e.MimeType_
}
func (e *entity) Description() string {
	return e.Description_
}
func (e *entity) Path() string {
	return e.Path_
}

type ImageEntity struct {
	entity `yaml:",inline"`
	Width  int `yaml:"width"`
	Height int `yaml:"height"`
}

type VideoEntity struct {
	entity   `yaml:",inline"`
	Width    int     `yaml:"width"`
	Height   int     `yaml:"height"`
	Duration float64 `yaml:"duraton"`
}

type AudioEntity struct {
	entity `yaml:",inline"`
}

File: backend/src/main/kotlin/net/hexe/the_gear_of_seasons/shelf/Entity.kt
package net.hexe.the_gear_of_seasons.shelf

import de.beosign.snakeyamlanno.property.YamlProperty
import org.yaml.snakeyaml.constructor.Constructor
import org.yaml.snakeyaml.introspector.Property
import org.yaml.snakeyaml.introspector.PropertyUtils
import java.nio.file.Paths
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.*
import kotlin.reflect.KClass


sealed class Entity {
  var dir: String = ""
  var id: String = ""
  var description: String = ""
  var date: Date = Date(0)

  @YamlProperty(key = "mime-type")
  var mimeType: String = "application/octet-stream"
  abstract fun metaFilename(): String
  abstract fun dataFilename(): String
  fun metaPath(): String = Paths.get(dir, metaFilename()).toAbsolutePath().toString()
  fun dataPath(): String = Paths.get(dir, dataFilename()).toAbsolutePath().toString()
  fun localTime(): LocalDateTime = LocalDateTime.ofInstant(date.toInstant(), ZoneId.systemDefault())
}

class Image : Entity() {
  companion object {
    const val kYamlExtension = ".image.yml"
  }
  var width: Int = 0
  var height: Int = 0

  override fun metaFilename(): String {
    return "$id$kYamlExtension"
  }
  override fun dataFilename(): String =
    when(mimeType) {
      "image/gif" -> "$id.gif"
      "image/jpeg" -> "$id.jpg"
      "image/png" -> "$id.png"
      else -> "$id.?"
    }
}

class Video : Entity() {
  companion object {
    const val kYamlExtension = ".video.yml"
  }
  var width: Int = 0
  var height: Int = 0
  var duraton: Float = 0.0f

  override fun metaFilename(): String {
    return "$id${kYamlExtension}"
  }
  override fun dataFilename(): String =
    when(mimeType) {
      "video/mp4" -> "$id.mp4"
      "video/x-matroska" -> "$id.mkv"
      else -> "$id.?"
    }
}

class Audio : Entity() {
  companion object {
    const val kYamlExtension = ".audio.yml"
  }

  override fun metaFilename(): String {
    return "$id${kYamlExtension}"
  }
  override fun dataFilename(): String =
    when(mimeType) {
      else -> "$id.?"
    }
}

class EntityPropertyUtils : PropertyUtils() {
  override fun getProperty(type: Class<out Any?>?, originalName: String): Property? {
    var name = originalName
    if (name.indexOf('-') > -1) {
      name = name.split('-').withIndex().map { (idx, word) ->
        if(idx == 0) {
          word
        } else {
          word[0].toUpperCase() + word.substring(1)
        }
      }.joinToString("")
    }
    return super.getProperty(type, name)
  }
}

fun <T : Entity> constructorOf(clazz: KClass<T>): Constructor {
  val c = Constructor(clazz.java)
  c.propertyUtils = EntityPropertyUtils()
  return c
}

